### PR TITLE
Execution type for multiple jobs in a phase: parallel or sequentially.

### PR DESCRIPTION
**Pull request by sshelomentsev**
Hi everyone,

I've implemented a feature that allows you to select the type of execution jobs in a phase.
This two options: parallel and sequentially. Running in parallel fully coincides with current execution logic. In the sequential case next job starts after the end of the previous one. The decision to launch the next job is based on phase configuration.

Hope, my request is clear.

screen shot 2016-02-12 at 17 11 21

### CODE REVIEW

**Comment by emoshaya-cognito** [comment]
Does this allow you to run multiple phases in parallel?

**Comment by sshelomentsev** [comment]
@emoshaya-cognito this feature just for multiple jobs in a phase

**Comment by emoshaya-cognito** [comment]
ok ... I was hoping to have the option for phases to also run in parallel. So, I'm guessing your change will introduce the ability to allow jobs within a phase to run sequentially as opposed to the default in parallel?

**Comment by sshelomentsev** [comment]
@emoshaya-cognito yes

**Comment by Brantone** [comment]
Isn't the purpose of this plugin to be that "jobs are parallel" and "phases are sequential" ? Or is there a confusion on how they're grouped and named? Otherwise seems somewhat redundant.

**Comment by seveneleven** [comment]
Does this feature preserve the given job order inside the job phase? Then, it's exactly what I've been looking for, 
In some phases, there are jobs that I need to execute sequentially (in a defined order) while I am now forced to create a separate job phase for each of them.

File: src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java [hunk 1]
```diff
@@ -102,10 +108,11 @@
 
     @DataBoundConstructor
     public MultiJobBuilder(String phaseName, List<PhaseJobsConfig> phaseJobs,
-            ContinuationCondition continuationCondition) {
+            ContinuationCondition continuationCondition, ExecutionType executionType) {
```
Comment (oleg-nenashev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: formally breaks the binary compatibility

File: src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java [hunk 2]
```diff
@@ -219,7 +226,7 @@ public boolean perform(final AbstractBuild<?, ? > build, final Launcher launcher
         Jenkins jenkins = Jenkins.getInstance();
         MultiJobBuild multiJobBuild = (MultiJobBuild) build;
         MultiJobProject thisProject = multiJobBuild.getProject();
-        Map<PhaseSubJob, PhaseJobsConfig> phaseSubJobs = new HashMap<PhaseSubJob, PhaseJobsConfig>(
+        Map<PhaseSubJob, PhaseJobsConfig> phaseSubJobs = new LinkedHashMap<PhaseSubJob, PhaseJobsConfig>(
```
Comment (oleg-nenashev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: Seems to be unrelated

File: src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java [hunk 3]
```diff
@@ -393,7 +413,7 @@ else if(!jobStatus.isBuildable() && !phaseConfig.isApplyConditionOnlyIfNoSCMChan
         return true;
     }
 
-    public final class SubJobWorker extends Thread {
+    public final class SubJobWorker implements Callable {
```
Comment (oleg-nenashev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: Now it consumes Jenkins workers instead of spawned threads. IMHO it's a good approach from the scalability PoV

Comment (oleg-nenashev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: It also breaks the binary compatibility BTW

File: src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java [hunk 4]
```diff
@@ -410,7 +430,7 @@ public SubJobWorker(MultiJobProject multiJobProject,
             this.queue = queue;
         }
 
-        public void run() {
+        public Object call() {
```
Comment (oleg-nenashev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: Makes sense to add Javadoc that the return value is useless. On the other hand, support of meaningful result object may be useful in the future (e.g. <code>SubJobCallResult</code>)

File: src/main/webapp/help-executionType.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Define how to run jobs in a phase: sequentially or parallel. Possibility to run jobs in parallel depends on the available executors.
```
Comment (oleg-nenashev) [inline src/main/webapp/help-executionType.html]: And Jenkins workers number...

File: src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java (hunk 2 above)
Comment (sshelomentsev) [inline src/main/java/com/tikal/jenkins/plugins/multijob/MultiJobBuilder.java]: Please explain what issue this raises? In code, set the default value that will be used for older job configurations and jobs will be executed in parallel.

**Comment by sshelomentsev** [comment]
Hi, @shacharbz @hagzag 
Have you a chance to review/merge this?
I think after the last update should not be problems with older versions of project configurations.
